### PR TITLE
Pecan-dev

### PR DESCRIPTION
**Pull request by shua1090**
# CAN PR
Contributors
- Shynn

## Overview
Adds CAN functionality (mostly pecan helper methods and abstract interfaces to platform-specific CAN libraries)

## Testing
Tested between Arduino and LPC 1768

## Future Work
- The ints should be platform-independent, but that was fixed in a later commit under pecan-sched
- Better documentation

### CODE REVIEW

File: src/common/pecan.hpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+struct CANPacket {
+    int id;
+    char data[8] = {0};
+    char dataSize = 0;
+};
+
+// 
+struct PCANListenParamsCollection{
+    CANListenParam arr[1000];
```
Comment (shua1090) [inline src/common/pecan.hpp:21]: Change to a lower number, we're not expecting to have so much ids to listen for

File: src/common/pecan.hpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    int size = 0;
+};
+
+// A single CANListenParam
+struct CANListenParam {
+    int listen_ids;
```
Comment (shua1090) [inline src/common/pecan.hpp:27]: listen_id, one per param

File: src/common/ptypes.hpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#ifndef PTYPES_H
+#define PTYPES_H
+
+#define MAX_SIZE_PACKET_DATA 8
```
Comment (shua1090) [inline src/common/ptypes.hpp:4]: is this used somewhere?

**Review by shua1090 — COMMENTED** [review]
Few minor fixes, check what to do with the unnecessary (testing) platform-specific code

**Comment by onlineknudsen** [comment]
Use lib_deps in platformio.ini file instead of git submodules

File: src/common/pecan.hpp (hunk 1 above)
Comment (onlineknudsen) [reply src/common/pecan.hpp:21]: magic number

File: src/common/pecan.hpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+/// @param listen_id the integer id we're listening for
+/// @param handler a function that takes in a
+///                pointer to the received packet and returns int
+///                for success/failure
+/// @return 0 on success, nonzero on Failure (see PCAN_ERR enum)
+int waitPacket(CANPacket * recv_pack, int listen_id, int (*handler)(CANPacket *));
+
+int sendPacket(CANPacket*  p);
+
+int getID(int fn_id, int node_id);
```
Comment (onlineknudsen) [inline src/common/pecan.hpp:45]: rename "combinedID"

File: src/common/ptypes.hpp [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+// DEFINITIONS, TYPEDEFS
+
+template <typename T>
```
Comment (onlineknudsen) [inline src/common/ptypes.hpp:7]: declare dont define

File: src/arch/arduino.cpp
Comment (onlineknudsen) [inline src/arch/arduino.cpp:1]: Test Arduino receive packet

File: src/common/pecan.hpp (hunk 1 above)
Comment (shua1090) [reply src/common/pecan.hpp]: done

File: src/common/pecan.hpp (hunk 2 above)
Comment (shua1090) [reply src/common/pecan.hpp:24]: done

File: src/common/pecan.hpp (hunk 4 above)
Comment (shua1090) [reply src/common/pecan.hpp]: done

File: src/common/ptypes.hpp (hunk 5 above)
Comment (shua1090) [reply src/common/ptypes.hpp:6]: done

File: src/arch/arduino.cpp
Comment (shua1090) [reply src/arch/arduino.cpp:1]: wip

**Comment by shua1090** [comment]
Testing example Telemetry (Arduino) -> MCU (LPC1768):
- Telemetry Code:
image
- MCU Code:
image
- Result:
image

**Comment by shua1090** [comment]
Testing example MCU -> Telemetry:
- MCU (Sending):
image
- Telemetry (Recv):
image

File: src/common/pecan.hpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+enum MATCH_TYPE{
+    MATCH_EXACT=0,
+    MATCH_SIMILAR=1,
+};
+
+// A CANPacket: takes in 11-bit id, 8 bytes of data
+struct CANPacket {
+    int id;
+    char data[MAX_SIZE_PACKET_DATA] = {0};
+    char dataSize = 0;
```
Comment (carterchannui) [inline src/common/pecan.hpp:27]: Perhaps `uint8_t` might be a better choice since it's more explicit and not platform dependent.

Comment (shua1090) [reply src/common/pecan.hpp:27]: but thank you for your feedback!!!

**Pull request closed without merging**